### PR TITLE
Add transaction solidifier queue

### PR DESCRIPTION
**Pull request by DyrellC**
# Description
Decouple the milestone and transaction solidification processes to handle transaction solidification asynchronously from milestone checking. Currently each time the `MilestoneSolidifier` checks the solidity of a transaction, it does the `TransactionSolidifier`'s `checkSolidity` call directly. This can cause a hanging when processing longer sync queues, which can delay the `MilestoneSolidifier` processing newer received milestone transactions. In order to decouple this, the `MilestoneSolidifier` will check the solid state of a transaction and add it to a processing queue in the `TransactionSolidifier` so that there are no hanging moments during milestone processing. This is a step introduced within the process of updating milestone handling as per #1674.

Follow up to #1805 
 
## Type of change
- Enhancement (a non-breaking change which adds functionality)

# How Has This Been Tested?
- All previously created tests still work, new unit tests work, and regression tests still pass locally (with the exception of the already known bug for machine3) 

# Checklist:
- [x] My code follows the style guidelines for this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: src/main/java/com/iota/iri/service/validation/impl/TransactionSolidifierImpl.java [hunk 1]
```diff
@@ -87,6 +104,43 @@ public void shutdown() {
...
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public boolean addMilestoneToSolidificationQueue(Hash hash){
+        try{
+            TransactionViewModel tx = fromHash(tangle, hash);
+            if(tx.isSolid()){
```
Comment (GalRogozinski) [inline src/main/java/com/iota/iri/service/validation/impl/TransactionSolidifierImpl.java:132]:
```suggestion
            if (tx.isSolid()) {
```

File: src/main/java/com/iota/iri/service/validation/impl/TransactionSolidifierImpl.java [hunk 2]
```diff
@@ -87,6 +104,43 @@ public void shutdown() {
...
+                transactionPropagator.addToPropagationQueue(hash);
+                return true;
+            }
+            addToSolidificationQueue(hash);
+            return false;
+        }catch(Exception e){
```
Comment (GalRogozinski) [inline src/main/java/com/iota/iri/service/validation/impl/TransactionSolidifierImpl.java:138]:
```suggestion
        } catch (Exception e) {
```

File: src/main/java/com/iota/iri/service/validation/impl/TransactionSolidifierImpl.java [hunk 3]
```diff
@@ -216,6 +280,12 @@ private void addToBroadcastQueue(TransactionViewModel tvm) {
         }
     }
 
+    @VisibleForTesting
+    Set<Hash> getSolidificationQueue(){
```
Comment (GalRogozinski) [inline src/main/java/com/iota/iri/service/validation/impl/TransactionSolidifierImpl.java:284]: It is funny that you call it `getQueue` but you return a `Set`

File: src/test/java/com/iota/iri/service/validation/impl/TransactionSolidifierImplTest.java [hunk 4]
```diff
@@ -98,11 +98,22 @@ public void verifyTxIsNotSolid() throws Exception {
         assertFalse("Expected transaction to fail solidity check", txSolidifier.checkSolidity(tx.getHash()));
     }
 
+    @Test
+    public void getSolidificationQueue() throws Exception {
```
Comment (GalRogozinski) [inline src/test/java/com/iota/iri/service/validation/impl/TransactionSolidifierImplTest.java:102]: You are testing `add` to solidification queue not `get`, so better rename

File: src/main/java/com/iota/iri/service/validation/impl/TransactionSolidifierImpl.java (hunk 3 above)
Comment (DyrellC) [reply src/main/java/com/iota/iri/service/validation/impl/TransactionSolidifierImpl.java:284]: A fair point, i'll rename

**Pull request merged**